core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/CsdLocationFilter.java

```diff
@@ -21,45 +20,71 @@
         if (StackType.WORKLOAD.equals(imageFilterParams.getStackType())) {
-            if (image == null || image.getPreWarmCsd() == null || image.getPreWarmCsd().isEmpty()) {
+            if (isInvalidImage(image)) {
                 LOGGER.debug("Image or CSD parcels are not present. Image: {}", image);
                 return false;
             } else {
-                List<String> csdList = getPreWarmCsdList(image, imageFilterParams);
-                LOGGER.debug("Matching URLs: [{}]", csdList);
-                return !csdList.isEmpty() && csdList.stream().allMatch(csdUrl -> URL_PATTERN.matcher(csdUrl).find());
+                Set<String> requiredParcels = getRequiredParcels(imageFilterParams);
+                LOGGER.debug("Stack related required parcels: {}", requiredParcels);
+                return requiredParcels.isEmpty() || isEligibleForUpgrade(image, requiredParcels);
             }
         } else {
             LOGGER.debug("Skip filtering because the stack type is {}", imageFilterParams.getStackType());
             return true;
         }
     }
 
-    private List<String> getPreWarmCsdList(Image image, ImageFilterParams imageFilterParams) {
-        Set<String> relatedParcelNames = getParcelNames(imageFilterParams);
-        List<String> imagePreWarmCsd = image.getPreWarmCsd();
-        return CollectionUtils.isEmpty(relatedParcelNames) ? imagePreWarmCsd
-                : filterPreWarmCsdListByStackRelatedParcels(relatedParcelNames, imagePreWarmCsd);
+    private boolean isInvalidImage(Image image) {
+        return image == null || image.getPreWarmCsd() == null || image.getPreWarmCsd().isEmpty();
     }
 
-    private List<String> filterPreWarmCsdListByStackRelatedParcels(Set<String> stackRelatedParcelNames, List<String> imagePreWarmCsd) {
-        return imagePreWarmCsd.stream()
-                .filter(preWarmCsd -> stackRelatedParcelNames.stream().anyMatch(csdUrlContainsParcelName(preWarmCsd)))
-                .collect(Collectors.toList());
+    private Set<String> getRequiredParcels(ImageFilterParams imageFilterParams) {
+        return Optional.ofNullable(imageFilterParams.getStackRelatedParcels())
+                .map(Map::keySet)
+                .orElse(Collections.emptySet())
+                .stream()
+                .filter(parcel -> !IGNORED_PARCEL_NAME.equals(parcel))
+                .collect(Collectors.toSet());
     }
 
-    private Set<String> getParcelNames(ImageFilterParams imageFilterParams) {
-        return Optional.ofNullable(imageFilterParams)
-                .map(ImageFilterParams::getStackRelatedParcels)
-                .map(Map::keySet)
-                .orElse(Collections.emptySet());
+    private boolean isEligibleForUpgrade(Image image, Set<String> requiredParcels) {
+        Map<String, Set<String>> csdUrlsByParcel = getCsdUrlsByParcelNames(image, requiredParcels);
+        LOGGER.debug("Available CSD URLs for the parcels: [{}]", csdUrlsByParcel);
```

Comment:
We should log the image uuid as well otherwise we won't know what this message refers to.